### ISSUE TITLE
failed assertions and uncaught exceptions when a receiver gets disconnected

### ISSUE THREAD

**Issue by rathann** [pwr-Solaar/Solaar#995]
**Information**
- Solaar version (`solaar --version` and `git describe --tags`): 1.0.4
- Distribution: Fedora 32
- Kernel version (ex. `uname -srmo`): Linux 5.8.16-200.fc32.x86_64 x86_64 GNU/Linux
- Output of `solaar show` for the target device (if applicable): N/A

**Describe the bug**
solaar sometimes fails with failed assertion or unhandled exception when the receiver gets disconnected in the middle of operation.

**To Reproduce**
1. Disconnect the receiver or the USB hub to which it's connected (or cut power to it).
2. Solaar sometimes fails with `AssertionError` or `NoReceiver` exception.

**Screenshots**
N/A

**Additional context**
Downstream bug reports:
* [`AssertionError`](https://bugzilla.redhat.com/show_bug.cgi?id=1891242) in [`tray.py:_update_menu_item`](https://github.com/pwr-Solaar/Solaar/blob/master/lib/solaar/ui/tray.py#L448)
* [`AssertionError`](https://bugzilla.redhat.com/show_bug.cgi?id=1891655) in [`window.py:_device_row`](https://github.com/pwr-Solaar/Solaar/blob/master/lib/solaar/ui/window.py#L435)
* [`NoReceiver` exception due to invalid file descriptor (`OSError`)](https://bugzilla.redhat.com/show_bug.cgi?id=1891717) in [`base.py:_skip_incoming`](https://github.com/pwr-Solaar/Solaar/blob/master/lib/logitech_receiver/base.py#L292) called all the way up from [`window.py:_device_selected`](https://github.com/pwr-Solaar/Solaar/blob/master/lib/solaar/ui/window.py#L397)

If you want, I can split these three into separate bug reports instead, but they seem to have a common theme of the receiver getting disconnected from the USB bus for one reason or another.

**Comment by rathann** [comment]
The `NoReceiver` bug report above is marked private, so let me repost the traceback here:

Version-Release number of selected component:
solaar-1.0.4-1.fc32

Additional info:
reporter:       libreport-2.13.1
cgroup:         0::/user.slice/user-1000.slice/user@1000.service/gnome-launched-solaar.desktop-2635.scope
cmdline:        /usr/bin/python3 -s /usr/bin/solaar --window=hide
crash_function: read
exception_type: OSError
executable:     /usr/bin/solaar
interpreter:    python3-3.8.6-1.fc32.x86_64
kernel:         5.8.16-200.fc32.x86_64
runlevel:       N 5
type:           Python3
uid:            1000

Truncated backtrace:
`base.py:292:_skip_incoming:logitech_receiver.base.NoReceiver: {'reason': OSError(9, 'Descrittore di file errato')}`

Traceback (most recent call last):
```
  File "/usr/lib/python3.8/site-packages/logitech_receiver/base.py", line 288, in _skip_incoming
    data = _hid.read(ihandle, _MAX_READ_SIZE, 0)
  File "/usr/lib/python3.8/site-packages/hidapi/udev.py", line 357, in read
    rlist, wlist, xlist = _select([device_handle], [], [device_handle], timeout)
OSError: [Errno 9] Descrittore di file errato

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/solaar/ui/window.py", line 397, in _device_selected
    _update_info_panel(device, full=True)
  File "/usr/lib/python3.8/site-packages/solaar/ui/window.py", line 792, in _update_info_panel
    _update_device_panel(device, _info._device, _info._buttons, full)
  File "/usr/lib/python3.8/site-packages/solaar/ui/window.py", line 673, in _update_device_panel
    device.status.read_battery(device)
  File "/usr/lib/python3.8/site-packages/logitech_receiver/status.py", line 251, in read_battery
    battery = _hidpp20.get_battery(d)
  File "/usr/lib/python3.8/site-packages/logitech_receiver/hidpp20.py", line 1131, in get_battery
    battery = feature_request(device, FEATURE.BATTERY_STATUS)
  File "/usr/lib/python3.8/site-packages/logitech_receiver/hidpp20.py", line 1016, in feature_request
    if device.online and device.features:
  File "/usr/lib/python3.8/site-packages/logitech_receiver/hidpp20.py", line 257, in _check
    reply = self.device.request(0x0000, _pack('!H', FEATURE.FEATURE_SET))
  File "/usr/lib/python3.8/site-packages/logitech_receiver/device.py", line 424, in request
    return _base.request(
  File "/usr/lib/python3.8/site-packages/logitech_receiver/base.py", line 384, in request
    _skip_incoming(handle, ihandle, notifications_hook)
  File "/usr/lib/python3.8/site-packages/logitech_receiver/base.py", line 292, in _skip_incoming
    raise NoReceiver(reason=reason)
logitech_receiver.base.NoReceiver: {'reason': OSError(9, 'Descrittore di file errato')}
```

Local variables in innermost frame:
handle: 16
ihandle: 16
notifications_hook: None

**Comment by pfps** [comment]
The first two error reports appear to be problems in the low-level code of Solaar.  Fixing them requires rewriting this code, which isn't going to happen quickly.  I've converted the assertions to warnings as it looks as if Solaar would continue reasonably under these conditions.

The last report appears to be a lack of recovery code when requesting information from a device and a disconnect happens.  I've added in the missing recovery code.

These all appear to be cases where a device or receiver disconnects when Solaar is doing something related to the device or receiver.  Verifying that the fixes work is difficult.  I've done a bit of testing but not comprehensive testing.
